File: generator/sources/csharp/KalturaClient/KalturaClientBase.cs
// ===================================================================================================
//                           _  __     _ _
//                          | |/ /__ _| | |_ _  _ _ _ __ _
//                          | ' </ _` | |  _| || | '_/ _` |
//                          |_|\_\__,_|_|\__|\_,_|_| \__,_|
//
// This file is part of the Kaltura Collaborative Media Suite which allows users
// to do with audio, video, and animation what Wiki platfroms allow them to do with
// text.
//
// Copyright (C) 2006-2011  Kaltura Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as
// published by the Free Software Foundation, either version 3 of the
// License, or (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.
//
// @ignore
// ===================================================================================================
using System;
using System.Collections.Generic;
using System.Text;
using System.Net;
using System.Web;
using System.IO;
using System.Security.Cryptography;
using System.Xml;
using System.Xml.XPath;
using System.Runtime.Serialization;
using System.Threading;

namespace Kaltura
{
    public class KalturaClientBase
    {
        #region Private Fields

        protected IDictionary<string, Object> clientConfiguration;
        protected IDictionary<string, Object> requestConfiguration;

        protected KalturaConfiguration _Config;
        private bool _ShouldLog;
        private List<KalturaServiceActionCall> _CallsQueue;
        private List<string> _MultiRequestReturnType;
        private KalturaParams _MultiRequestParamsMap;
        private WebHeaderCollection _ResponseHeaders;

        #endregion

        #region Properties

        public bool IsMultiRequest
        {
            get { return (_MultiRequestReturnType != null); }
        }

        public WebHeaderCollection ResponseHeaders
        {
            get { return _ResponseHeaders; }
        }

        #endregion

        #region CTor

        public KalturaClientBase(KalturaConfiguration config)
        {
            clientConfiguration = new Dictionary<string, Object>();
            requestConfiguration = new Dictionary<string, Object>();

            _Config = config;
            if (_Config.Logger != null)
            {
                _ShouldLog = true;
            }
            _CallsQueue = new List<KalturaServiceActionCall>();
            _MultiRequestParamsMap = new KalturaParams();
        }

        #endregion

        #region Methods

        public void QueueServiceCall(string service, string action, string fallbackClass, KalturaParams kparams)
        {
            this.QueueServiceCall(service, action, fallbackClass, kparams, new KalturaFiles());
        }

        public void QueueServiceCall(string service, string action, string fallbackClass, KalturaParams kparams, KalturaFiles kfiles)
        {
            Object value;
            foreach (string param in requestConfiguration.Keys)
            {
                value = requestConfiguration[param];
                if (value is KalturaObjectBase)
                {
                    kparams.Add(param, ((KalturaObjectBase)value).ToParams());
                }
                else
                {
                    kparams.Add(param, value.ToString());
                }
            }

            KalturaServiceActionCall call = new KalturaServiceActionCall(service, action, kparams, kfiles);
            if (_MultiRequestReturnType != null)
                _MultiRequestReturnType.Add(fallbackClass);
            this._CallsQueue.Add(call);
        }

        public XmlElement DoQueue()
        {
            if (_CallsQueue.Count == 0)
            {
                resetRequest();
                return null;
            }

            DateTime startTime = DateTime.Now;

            this.Log("service url: [" + this._Config.ServiceUrl + "]");

            KalturaParams kparams = new KalturaParams();
            KalturaFiles kfiles = new KalturaFiles();

            foreach (string param in clientConfiguration.Keys)
            {
                kparams.Add(param, clientConfiguration[param].ToString());
            }
            kparams.AddIfNotNull("format", this._Config.ServiceFormat.GetHashCode());

            string url = this._Config.ServiceUrl + "/api_v3";

            if (_MultiRequestReturnType != null)
            {
                url += "/service/multirequest";
                int i = 1;
                foreach (KalturaServiceActionCall call in _CallsQueue)
                {
                    KalturaParams callParams = call.GetParamsForMultiRequest(i);
                    kparams.Add(callParams);
                    KalturaFiles callFiles = call.GetFilesForMultiRequest(i);
                    kfiles.Add(callFiles);
                    i++;
                }

                // map params
                foreach (KeyValuePair<string, IKalturaSerializable> item in _MultiRequestParamsMap)
                {
                    string requestParam = item.Key;
                    IKalturaSerializable resultParam = item.Value;

                    if (kparams.ContainsKey(requestParam))
                    {
                        kparams[requestParam] = resultParam;
                    }
                }
            }
            else
            {
                KalturaServiceActionCall call = _CallsQueue[0];
                url += "/service/" + call.Service + "/action/" + call.Action;
                kparams.Add(call.Params);
                kfiles.Add(call.Files);
            }

            kparams.Add("kalsig", this.Signature(kparams));
            string json = kparams.ToJson();

            this.Log("full reqeust url: [" + url + "]");
            this.Log("full reqeust data: [" + json + "]");

            // build request
            HttpWebRequest request = (HttpWebRequest)HttpWebRequest.Create(url);
            if (kfiles.Count == 0)
            {
                request.Timeout = _Config.Timeout;
            }
            else
            {
                request.Timeout = Timeout.Infinite;
            }
            request.Method = "POST";
            request.AutomaticDecompression = DecompressionMethods.GZip | DecompressionMethods.Deflate;
            request.Headers = _Config.RequestHeaders;
            request.Accept = "application/json";

            // Add proxy information if required
            createProxy(request, _Config);

            if (kfiles.Count > 0)
            {
                this.PostMultiPartWithFiles(request, json, kfiles);
            }
            else
            {
                this.PostUrlEncodedParams(request, json);
            }

            // get the response
            using (WebResponse response = request.GetResponse())
            {
                Encoding enc = System.Text.Encoding.UTF8;
                StreamReader responseStream = new StreamReader(response.GetResponseStream(), enc);
                string responseString = responseStream.ReadToEnd();

                this._ResponseHeaders = response.Headers;
                string serverName = null;
                string serverSession = null;
                for (int i = 0; i < this._ResponseHeaders.Count; ++i)
                {
                    if (this._ResponseHeaders.Keys[i] == "X-Me")
                        serverName = this._ResponseHeaders[i];
                    if (this._ResponseHeaders.Keys[i] == "X-Kaltura-Session")
                        serverSession = this._ResponseHeaders[i];
                }
                if (serverName != null || serverSession != null)
                    this.Log("server: [" + serverName + "], session: [" + serverSession + "]");

                this.Log("result (serialized): " + responseString);

                DateTime endTime = DateTime.Now;

                this.Log("execution time for [" + url + "]: [" + (endTime - startTime).ToString() + "]");

                XmlDocument xml = new XmlDocument();
                xml.LoadXml(responseString);

                this.ValidateXmlResult(xml);
                XmlElement result = xml["xml"]["result"];
                this.ThrowExceptionOnAPIError(result);

                if(!IsMultiRequest)
                    resetRequest();

                return result;
            }
        }
        private void createProxy(HttpWebRequest request, KalturaConfiguration _Config)
        {
            if (String.IsNullOrEmpty(_Config.ProxyAddress))
                return;
            Console.WriteLine("Create proxy");
            if (!(String.IsNullOrEmpty(_Config.ProxyUser) || String.IsNullOrEmpty(_Config.ProxyPassword)))
            {
                ICredentials credentials = new NetworkCredential(_Config.ProxyUser, _Config.ProxyPassword);
                request.Proxy = new WebProxy(_Config.ProxyAddress, false, null, credentials);
            }
            else
            {
                request.Proxy = new WebProxy(_Config.ProxyAddress);
            }
        }

        public void StartMultiRequest()
        {
            _MultiRequestReturnType = new List<string>();
        }

        public KalturaMultiResponse DoMultiRequest()
        {
            XmlElement multiRequestResult = DoQueue();

            KalturaMultiResponse multiResponse = new KalturaMultiResponse();
            if (multiRequestResult == null)
            {
                resetRequest();
                return multiResponse;
            }
            
            multiResponse = ParseMultiRequestResult(multiRequestResult);
            resetRequest();
            return multiResponse;
        }
        
        private KalturaMultiResponse ParseMultiRequestResult(XmlElement childNode) 
        {
            return ParseMultiRequestResult(childNode, true);
        }
        
        private KalturaMultiResponse ParseMultiRequestResult(XmlElement childNode, bool incrementI)  
        {
            int i = 0;
            KalturaMultiResponse multiResponse = new KalturaMultiResponse();
            foreach(XmlElement arrayNode in childNode.ChildNodes) 
            {
                XmlElement error = arrayNode["error"];
                if (error != null && error["code"] != null && error["message"] != null) 
                    multiResponse.Add(new KalturaAPIException(error["code"].InnerText, error["message"].InnerText));
                else if (arrayNode["objectType"] != null)
                    multiResponse.Add(KalturaObjectFactory.Create(arrayNode, _MultiRequestReturnType[i]));
                else if (arrayNode["item"] != null)
                    multiResponse.Add(ParseMultiRequestResult(arrayNode, false));
                else
                    multiResponse.Add(arrayNode.InnerText);
                if (incrementI) 
                    i++;
            }
    
            return multiResponse;
        }
        
        public void MapMultiRequestParam(int resultNumber, int requestNumber, string requestParamName)
        {
            this.MapMultiRequestParam(resultNumber, null, requestNumber, requestParamName);
        }

        public void MapMultiRequestParam(int resultNumber, string resultParamName, int requestNumber, string requestParamName)
        {
            string resultParam = "{" + resultNumber + ":result";
            if (resultParamName != null && resultParamName != "")
                resultParam += resultParamName;
            resultParam += "}";

            string requestParam = requestNumber + ":" + requestParamName;

            _MultiRequestParamsMap.Add(requestParam, resultParam);
        }

        public string GenerateSession(string adminSecretForSigning)
        {
            return this.GenerateSession(adminSecretForSigning, "");
        }

        public string GenerateSession(string adminSecretForSigning, string userId)
        {
            return this.GenerateSession(adminSecretForSigning, userId, (KalturaSessionType)(0));
        }

        public string GenerateSession(string adminSecretForSigning, string userId, KalturaSessionType type)
        {
            return this.GenerateSession(adminSecretForSigning, userId, type, -1);
        }

        public string GenerateSession(string adminSecretForSigning, string userId, KalturaSessionType type, int partnerId)
        {
            return this.GenerateSession(adminSecretForSigning, userId, type, partnerId, 86400);
        }

        public string GenerateSession(string adminSecretForSigning, string userId, KalturaSessionType type, int partnerId, int expiry)
        {
            return this.GenerateSession(adminSecretForSigning, userId, type, partnerId, expiry, "");
        }

        public string GenerateSession(string adminSecretForSigning, string userId, KalturaSessionType type, int partnerId, int expiry, string privileges)
        {
            string ks = string.Format("{0};{0};{1};{2};{3};{4};{5};", partnerId, UnixTimeNow() + expiry, type.GetHashCode(), DateTime.Now.Ticks, userId, privileges);

            SHA1 sha = new SHA1CryptoServiceProvider();

            byte[] ksTextBytes = Encoding.ASCII.GetBytes(adminSecretForSigning + ks);

            byte[] sha1Bytes = sha.ComputeHash(ksTextBytes);

            string sha1Hex = "";
            foreach (char c in sha1Bytes)
                sha1Hex += string.Format("{0:x2}", (int)c);

            ks = sha1Hex.ToLower() + "|" + ks;

            return EncodeTo64(ks);
        }

        #endregion

        #region Private Helpers

        private void Log(string msg)
        {
            if (this._ShouldLog)
            {
                this._Config.Logger.Log(msg);
            }
        }

        private string Signature(KalturaParams kparams)
        {
            MD5CryptoServiceProvider md5 = new MD5CryptoServiceProvider();
            byte[] data = Encoding.ASCII.GetBytes(kparams.ToJson());
            data = md5.ComputeHash(data);
            StringBuilder sBuilder = new StringBuilder();
            for (int i = 0; i < data.Length; i++)
            {
                sBuilder.Append(data[i].ToString("x2"));
            }
            return sBuilder.ToString();
        }

        private void ValidateXmlResult(XmlDocument doc)
        {
            XmlElement xml = doc["xml"];
            if (xml != null)
            {
                XmlElement result = xml["result"];
                if (result != null)
                {
                    return;
                }
            }

            throw new SerializationException("Invalid result");
        }

        private void ThrowExceptionOnAPIError(XmlElement result)
        {
            XmlElement error = result["error"];
            if (error != null && error["code"] != null && error["message"] != null)
                throw new KalturaAPIException(error["code"].InnerText, error["message"].InnerText);
        }

        private void PostMultiPartWithFiles(HttpWebRequest request, string json, KalturaFiles kfiles)
        {
            string boundary = "---------------------------" + DateTime.Now.Ticks.ToString("x");
            request.ContentType = "multipart/form-data; boundary=" + boundary;

            byte[] paramsBuffer = BuildMultiPartParamsBuffer(json, boundary);

            SortedList<string, MultiPartFileDescriptor> filesDescriptions = new SortedList<string, MultiPartFileDescriptor>();
            foreach (KeyValuePair<string, Stream> file in kfiles)
                filesDescriptions.Add(file.Key, BuildMultiPartFileDescriptor(file, boundary));

            // Set content's length
            request.ContentLength = paramsBuffer.LongLength;
            foreach (KeyValuePair<string, MultiPartFileDescriptor> fileDesc in filesDescriptions)
                request.ContentLength += fileDesc.Value.GetTotalLength();

            // And let's upload
            request.AllowWriteStreamBuffering = false; // A more sensible approach may be relevant
            Stream requestStream = request.GetRequestStream();
            requestStream.Write(paramsBuffer, 0, paramsBuffer.Length);
            foreach (KeyValuePair<string, MultiPartFileDescriptor> fileDesc in filesDescriptions)
            {
                requestStream.Write(fileDesc.Value._Header, 0, fileDesc.Value._Header.Length);

                byte[] buffer = new Byte[checked(Math.Min((uint)4096, fileDesc.Value._Stream.Length))];
                int bytesRead = 0;
                while ((bytesRead = fileDesc.Value._Stream.Read(buffer, 0, buffer.Length)) != 0)
                    requestStream.Write(buffer, 0, bytesRead);

                requestStream.Write(fileDesc.Value._Footer, 0, fileDesc.Value._Footer.Length);
            }

            requestStream.Close();
        }

        /// <summary>
        /// Compiles the parameters required for PostMultiPartWithFiles(...) into a byte buffer ready to be streamed.
        /// </summary>
        /// <param name="kparams">The request's parameters.</param>
        /// <param name="boundary">The multipart's boundary.</param>
        /// <returns>
        /// The parameters' byte array ready to be streamed.
        /// </returns>
        private byte[] BuildMultiPartParamsBuffer(string json, string boundary)
        {
            StringBuilder sb = new StringBuilder();
            sb.Append("--" + boundary + "\r\n");
            sb.Append("Content-Disposition: form-data; name=\"json\"\r\n");
            sb.Append("\r\n");
            sb.Append(HttpUtility.UrlDecode(json));
            sb.Append("\r\n--" + boundary + "\r\n");

            return Encoding.UTF8.GetBytes(sb.ToString()); 
        }

        /// <summary>
        /// Pre-compiles the multipart data required for a given file.
        /// </summary>
        /// <param name="fileEntry">The provided file infos.</param>
        /// <param name="boundary">The multipart's boundary.</param>
        /// <returns>
        /// A description containing the file's stream and the part's header and footer.
        /// </returns>
        private MultiPartFileDescriptor BuildMultiPartFileDescriptor(KeyValuePair<string, Stream> fileEntry, string boundary)
        {
            MultiPartFileDescriptor result = new MultiPartFileDescriptor();
            result._Stream = fileEntry.Value;

            // Build header
            StringBuilder sb = new StringBuilder();
            Stream fileStream = fileEntry.Value;
            if(fileStream is FileStream)
            {
                FileStream fs = (FileStream)fileStream;
                sb.Append("Content-Disposition: form-data; name=\"" + fileEntry.Key + "\"; filename=\"" + Path.GetFileName(fs.Name) + "\"" + "\r\n");
            }
            else if(fileStream is MemoryStream)
            {
                sb.Append("Content-Disposition: form-data; name=\"" + fileEntry.Key + "\"; filename=\"Memory-Stream-Upload\"" + "\r\n");
            }
            sb.Append("Content-Type: application/octet-stream" + "\r\n");
            sb.Append("\r\n");
            result._Header = Encoding.UTF8.GetBytes(sb.ToString());

            result._Footer = Encoding.UTF8.GetBytes("\r\n--" + boundary + "\r\n");

            return result;
        }

        private void PostUrlEncodedParams(HttpWebRequest request, string json)
        {
            byte[] buffer;
            buffer = System.Text.Encoding.UTF8.GetBytes(json);
            request.ContentType = "application/json";
            request.ContentLength = json.Length;
            Stream requestStream = request.GetRequestStream();
            requestStream.Write(buffer, 0, buffer.Length);
            requestStream.Close();
        }

        public long UnixTimeNow()
        {
            TimeSpan _TimeSpan = (DateTime.UtcNow - new DateTime(1970, 1, 1, 0, 0, 0));
            return (long)_TimeSpan.TotalSeconds;
        }

        private string EncodeTo64(string toEncode)
        {
            byte[] toEncodeAsBytes = System.Text.ASCIIEncoding.ASCII.GetBytes(toEncode);
            string returnValue = System.Convert.ToBase64String(toEncodeAsBytes);
            return returnValue;
        }

        protected void resetRequest()
        {
            if (IsMultiRequest)
            {
                _MultiRequestReturnType.Clear();
                _MultiRequestReturnType = null;
            }

            _CallsQueue.Clear();
            _MultiRequestParamsMap.Clear();
        }

        #endregion

        #region Support Types

        private struct MultiPartFileDescriptor
        {
            public Stream _Stream;
            public byte[] _Header;
            public byte[] _Footer;

            public long GetTotalLength()
            {
                return _Stream.Length + _Header.LongLength + _Footer.LongLength;
            }
        }

        #endregion
    }
}

File: generator/sources/ruby/lib/kaltura_client_base.rb
# ===================================================================================================
#                           _  __     _ _
#                          | |/ /__ _| | |_ _  _ _ _ __ _
#                          | ' </ _` | |  _| || | '_/ _` |
#                          |_|\_\__,_|_|\__|\_,_|_| \__,_|
#
# This file is part of the Kaltura Collaborative Media Suite which allows users
# to do with audio, video, and animation what Wiki platfroms allow them to do with
# text.
#
# Copyright (C) 2006-2011  Kaltura Inc.
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU Affero General Public License as
# published by the Free Software Foundation, either version 3 of the
# License, or (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Affero General Public License for more details.
#
# You should have received a copy of the GNU Affero General Public License
# along with this program.  If not, see <http:#www.gnu.org/licenses/>.
#
# @ignore
# ===================================================================================================
require 'rubygems'
require 'json'
require 'net/http'
require 'digest/md5'
require 'rexml/document'
require 'rest-client'
require 'openssl'
require 'digest/sha1'
require 'base64'
require 'date'
require 'yaml'

module Kaltura
	class KalturaNotImplemented; end

	class KalturaClientBase
		attr_accessor 	:config
		attr_reader 	:is_multirequest
		attr_reader 	:responseHeaders
		def initialize(config)
			@should_log = false
			@config = config
			@calls_queue = []
			@client_configuration = {}
			@request_configuration = {}

			if @config.logger != nil
				@should_log = true
			end
		end

		def queue_service_action_call(service, action, return_type, params = {}, files = {})
			# in start session partner id is optional (default nil). if partner id was not set, use the one in the config

			@request_configuration.each do |key, value|
				add_param(params, key, value)
			end

			call = KalturaServiceActionCall.new(service, action, return_type, params, files);
			@calls_queue.push(call);
		end

		def do_queue()
			begin
				@responseHeaders = {}
				start_time = Time.now

				if @calls_queue.length == 0
					@is_multirequest = false
					return []
				end

				log('service url: [' + @config.service_url + ']')

				# append the basic params
				params = {}
				files = {}
					
				url = @config.service_url+"/api_v3/"
				if (@is_multirequest)
					url += "service/multirequest/"
					i = 0
					@calls_queue.each do |call|
						call_params = call.get_params_for_multirequest(i)
						params.merge!(call_params)
						call_files = call.get_files_for_multirequest(i)
						files.merge!(call_files)
						i = i.next
					end
				else
					call = @calls_queue[0]
					url += "service/#{call.service}/action/#{call.action}/"
					params.merge!(call.params)
					files = call.files
				end
				
				add_param(params, "format", @config.format)
				@client_configuration.each do |key, value|
					add_param(params, key, value)
				end

				signature = signature(params)
				add_param(params, "kalsig", signature)

				log("url: " + url)
				log("params: " + params.to_yaml)

				result = do_http_request(url, params, files)

				@responseHeaders = result.headers
				log("server: [" + result.headers[:x_me].to_s + "], session: [" + result.headers[:x_kaltura_session].to_s + "]")

				log("result (xml): " + result.body)

				result_object = parse_to_objects(result.body)

				# reset
				reset_request()

				log("result (object yaml dump): " + result_object.to_yaml)

				end_time = Time.now

				log("execution time for [#{url}]: [#{end_time - start_time}]")

				return result_object

			rescue KalturaAPIError => e
				raise e
			rescue Exception => e
				raise KalturaAPIError.new("KALTURA_RUBY_CLIENT_ERROR", e.to_s)
			end
		end

		def reset_request()
			@calls_queue = []
			@is_multirequest = false
		end

		def get_serve_url()
			url = @config.service_url+"/api_v3/service/"

			call = @calls_queue[0]
			url += call.service + "/action/" + call.action
			params = call.params

			# reset
			@calls_queue = []
			@is_multirequest = false

			query_string = '?'
			params.each do |name, value|
				query_string << "#{name}=#{CGI::escape(value.to_s)}&"
			end

			serve_url = "#{url}#{query_string}"

			log("serve_url: " + serve_url)

			return serve_url
		end

		def do_http_request(url, params, files)

			headers = @config.requestHeaders
			headers[ 'Accept'] = 'text/xml';
				
			payload = {}
				
			if(files.size > 0)
				payload = files
				payload[:json] = params.to_json
			else
				payload = params.to_json
				headers[ 'Content-Type'] = 'application/json';
			end
			
			options = {
				:method => :post, 
				:url => url, 
				:headers => headers,
				:timeout => @config.timeout,
				:open_timeout => @config.timeout,
				:payload => payload
			}

			log("request options: " + JSON.pretty_generate(options))
			res = RestClient::Request.execute(options)

			return res
		end

		def self.object_from_xml(xml_element, return_type = nil)
			instance = nil
			if xml_element.elements.size > 0
				if xml_element.elements[1].name == 'item' # array or map
					if (xml_element.elements[1].elements['itemKey'].nil?) # array
						instance = []
						xml_element.elements.each('item') do | element |
							instance.push(KalturaClientBase.object_from_xml(element, return_type))
						end
					else # map
						instance = {}
						xml_element.elements.each('item') do | element |
							item_key = element.get_text('itemKey').to_s
							instance[item_key] = KalturaClientBase.object_from_xml(element, return_type)
						end
					end
				else # object
					object_type_element = xml_element.get_text('objectType')
					if (object_type_element != nil)
						object_class = xml_element.get_text('objectType').value.to_s
						kalturaModule = Module.const_get("Kaltura")
						
						begin
							instance = kalturaModule.const_get(object_class).new
						rescue NameError => e
							if(return_type != nil)
								instance = kalturaModule.const_get(return_type).new
							else
								raise e
							end
						end

						instance.from_xml(xml_element);
					else # error
						error_element = xml_element.elements['error']
						if (error_element != nil)
							code = xml_element.elements["error/code"].text
							message = xml_element.elements["error/message"].text

							instance = KalturaAPIError.new(code, message)
						end
					end
				end
			elsif return_type == nil
				return nil 
			else # simple type
				value = xml_element.text
				if return_type == "int"
					return value.to_i
				end
				
				return value
			end

			return instance;
		end

		def self.camelcase_to_underscore(val)
			val.gsub(/(.)([A-Z])/,'\1_\2').downcase
		end

		def parse_to_objects(data)
			parse_xml_to_objects(data)
		end

		def parse_xml_to_objects(xml)
			doc = REXML::Document.new(xml)
			raise_exception_if_error(doc)
			if (@is_multirequest)
				results = {}
				request_index = 0
				doc.elements.each('xml/result/*') do | element |
					results[request_index] = KalturaClientBase.object_from_xml(element, @calls_queue[request_index].return_type)
					request_index += 1
				end
				return results
			else
				doc.elements.each('xml/result') do | element |
					return KalturaClientBase.object_from_xml(element, @calls_queue.first.return_type)
				end
			end
		end

		def raise_exception_if_error(doc)
			if is_error(doc)
				code = doc.elements["xml/result/error/code"].text
				message = doc.elements["xml/result/error/message"].text
				raise KalturaAPIError.new(code, message)
			end
		end

		def is_error(doc)
			return doc.elements["xml/result/error/message"] && doc.elements["xml/result/error/code"];
		end

		def start_multirequest()
			@is_multirequest = true
		end

		def do_multirequest()
			return do_queue()
		end

		def signature(params)
			kParams = params.select { |key, value| !value.is_a?(File) }
			str = kParams.keys.map {|key| key.to_s }.sort.map {|key|
				"#{key}#{params[key]}"
			}.join("")

			Digest::MD5.hexdigest(str)
		end

		def add_param(params, name, value)
			if value == KalturaNotImplemented
				return
			elsif value == nil
				params[name + '__null'] = ''
			elsif value.is_a? Hash
				params[name] = {}
				if value.empty?
					add_param(params[name], "-", "");
				else
					value.each do |sub_name, sub_value|
						add_param(params[name], sub_name, sub_value);
					end
				end
			elsif value.is_a? Array
				if value.empty?
					params[name] = {}
					add_param(params[name], "-", "");
				else
					params[name] = Array.new(value.size)
					value.each_with_index do |ele, i|
						if ele.is_a? KalturaObjectBase
							add_param(params[name], i, ele.to_params)
						end
					end
				end
			elsif value.is_a? KalturaObjectBase
				add_param(params, name, value.to_params)
			else
				params[name] = value
			end
		end

		# Escapes a query parameter. Taken from RFuzz
		def escape(s)
			s.to_s.gsub(/([^ a-zA-Z0-9_.-]+)/n) {
				'%' + $1.unpack('H2'*$1.size).join('%').upcase
			}.tr(' ', '+')
		end

		def log(msg)
			if @should_log
				config.logger.log(Logger::INFO, msg)
			end
		end

		def generate_session(admin_secret, user_id, kaltura_session_type, partner_id, expiry=86400, privileges=nil)

			session = "#{partner_id};#{partner_id};#{Time.now.to_i + expiry};#{kaltura_session_type};#{rand.to_s.gsub("0.", "")};#{user_id};#{privileges};"

			digest_generator = OpenSSL::Digest.new('sha1')

			digest_generator.update(admin_secret)
			digest_generator.update(session)

			digest = digest_generator.hexdigest

			signature = digest + "|" + session
			b64 = Base64.encode64(signature)
			cleaned = b64.gsub("\n","")

			self.ks = cleaned
		end
	end

	class KalturaServiceActionCall
		attr_accessor :service
		attr_accessor :action
		attr_accessor :return_type
		attr_accessor :params
		attr_accessor :files
		
		def initialize(service, action, return_type, params, files)
			@service = service
			@action = action
			@return_type = return_type
			@params = parse_params(params)
			@files = files
		end

		def parse_params(params)
			new_params = {}
			params.each do |key, val|
				if val.kind_of? Hash
					new_params[key] = parse_params(val)
				else
					new_params[key] = val
				end
			end
			return new_params
		end

		def get_params_for_multirequest(multirequest_index)
			multirequest_params = {}
			multirequest_params[multirequest_index.to_s] = {}
			multirequest_params[multirequest_index.to_s]["service"] = @service
			multirequest_params[multirequest_index.to_s]["action"] = @action
			@params.each do |key, value|
				multirequest_params[multirequest_index.to_s][key] = value
			end
			return multirequest_params
		end
		
		def get_files_for_multirequest(multirequest_index)
			multirequest_params = {}
			@files.each do |key, value|
				multirequest_params[multirequest_index.to_s + ":" + key] = value
			end
			return multirequest_params
		end
	end

	class KalturaObjectBase
		attr_accessor :object_type
		attr_accessor :related_objects

		def from_xml(xml_element)
			self.related_objects = KalturaClientBase.object_from_xml(xml_element.elements['relatedObjects'], 'KalturaListResponse')
		end
		
		def to_params
			params = {};
			params["objectType"] = self.class.name.split('::').last
			instance_variables.each do |var|
				value = instance_variable_get(var)
				var = var.to_s.sub('@', '')
				kvar = camelcase(var)
				if (value != nil)
					if (value.is_a? KalturaObjectBase)
						params[kvar] = value.to_params;
					else
						params[kvar] = value;
					end
				else
					params[kvar] = value;
				end
			end
			return params;
		end

		def to_b(val)
			return [true, 'true', 1, '1'].include?(val.is_a?(String) ? val.downcase : val)
		end

		def camelcase(val)
			val = val.split('_').map { |e| e.capitalize }.join()
			val[0,1].downcase + val[1,val.length]
		end
	end

	class KalturaServiceBase
		attr_accessor :client
		def initialize(client)
			@client = client
		end
	end

	class KalturaConfiguration
		attr_accessor :logger
		attr_accessor :service_url
		attr_accessor :format
		attr_accessor :timeout
		attr_accessor :requestHeaders
		#
		# Adding service_url to the initialize signature to pass url to your own kaltura ce instance
		# Default is still set to http://www.kaltura.com.
		#
		def initialize()
			@service_url 	= "http://www.kaltura.com"
			@format 		= 2 # xml
			@timeout 		= 120
			@requestHeaders = {}
		end

		def service_url=(url)
			@service_url = url.chomp('/')
		end
	end

	class KalturaAPIError < RuntimeError
		attr_reader :code
		attr_reader :message
		def initialize(code, message)
			@code = code
			@message = message
		end
	end
end